src/firebase_functions/options.py

```diff
@@ -223,32 +197,141 @@ def asdict_with_global_options(self) -> dict:
                 merged_options[key] = global_options[key]
         # None values are automatically stripped out in ManifestEndpoint generation.
 
-        # _Sentinel values are converted to `None` in ManifestEndpoint generation
+        # _util.Sentinel values are converted to `None` in ManifestEndpoint generation
         # after other None values are removed - so as to keep them in the generated
         # YAML output as 'null' values.
         return merged_options
 
-
-@_dataclasses.dataclass(frozen=True)
-class _PubSubOptions(_ProviderOptions):
+    def _endpoint(self, **kwargs) -> _manifest.ManifestEndpoint:
+        assert kwargs["func_name"] is not None
+        options_dict = self._asdict_with_global_options()
+        options = self.__class__(**options_dict)
+
+        secret_envs: list[
+            _manifest.SecretEnvironmentVariable] | _util.Sentinel = []
+        if options.secrets is not None:
+            if isinstance(options.secrets, list):
+
+                def convert_secret(
+                        secret) -> _manifest.SecretEnvironmentVariable:
+                    secret_value = secret
+                    if isinstance(secret, SecretParam):
+                        secret_value = secret.name
+                    return {"key": secret_value}
+
+                secret_envs = list(
+                    map(convert_secret, _typing.cast(list, options.secrets)))
+            elif options.secrets is _util.Sentinel:
+                secret_envs = _typing.cast(_util.Sentinel, options.secrets)
+
+        region: list[str] | None = None
+        if isinstance(options.region, list):
```

Comment:
subtlety note: we allow region to be a list for HTTPS functions but no other functions. Yet in the functions.yaml it will always be a list for simplicity.